rfcs/20200504-tff-aggregators-placement.md

```diff
@@ -0,0 +1,163 @@
+model, which in a federated context generally assumes an additional
+party to perform the secure shuffling needed to realize a differential privacy
+guarantee. Since this is an established area of the literature with strong
+motivations and results, we see this as an important line of work for TFF to
+support in order to keep with its
+[project goals](https://github.com/tensorflow/federated#tensorflow-federated).
+
+In general, any secure aggregation protocol can be represented as a coordinated
+computation between three groups of parties: a server, a (potentially singleton)
+set of aggregators, and a set of clients. Note that these need not be mutually
+exclusive, so for example the traditional parameter server setting can be recovered
+as a special case by treating the server as a singleton aggregators set.
+
+The TFF Federated Core (FC) language currently realizes logically-distinct parties
+as "placements". While there exist `tff.SERVER` and `tff.CLIENTS` placements in FC,
+there is no `tff.AGGREGATORS` placement. Without such a placement, implementing new
+aggregation protocols in TFF can require low-level programming of the TFF executor
+stacks, as evidenced by
+[this community attempt to integrate secure aggregation](https://github.com/tf-encrypted/rfcs/blob/master/20190924-tensorflow-federated/integration-strategies.md).
+By adding a new `tff.AGGREGATORS` placement, users can more easily implement new
+aggregation protocols by expressing them as federated computations in FC.
+
+## User Benefit
+
+*How will users (or other contributors) benefit from this work? What would be the
+headline in the release notes or blog post?*
+
+Users can now express custom aggregation protocols in the Federated Core by working
+with federated data placed on `tff.AGGREGATORS`. Users will be unencumbered by the
+constraints of the current federated types in FC.
+
+## Design Proposal
+
+*This is the meat of the document, where you explain your proposal. If you have
+multiple alternatives, be sure to use sub-sections for better separation of the
+idea, and list pros/cons to each approach. If there are alternatives that you
+have eliminated, you should also list those here, and explain why you believe
+your chosen approach is superior.
+
+Adding the `tff.AGGREGATORS` placement for federated types involves adding a new `Placement` and `PlacementLiteral`, and then extending the compiler to recognize federated values with this placement when computing intrinsics. The compiler generally defines separate intrinsics by placement; e.g., `tff.federated_value(value, placement)` is actually interpreted by the compiler as `federated_value_at_clients(value)` or `federated_value_at_server(value)`, depending on the provided `placement`. This means we we will want to add new intrinsics that correspond to `tff.AGGREGATORS`, e.g. `federated_value_at_aggregators`.
+
+Existing federated computation that will need modification fall into the two categories below:
```

Comment:
It is not clear to me that we need to include non-federated intrinsics in this list? sequence_sum/reduce

Reply:
I took a deeper look, and you're likely right about not needing to modify the IntrinsicDefs for these.  But their factory functions will need to be adjusted to account for the new placement due to checks like [this one](https://github.com/tensorflow/federated/blob/c671f0b8857a3be18208db7a1c460f51650a7c60/tensorflow_federated/python/core/impl/intrinsic_factory.py#L366).

Reply:
The main reason I included them is that they need to be able to handle federated values that have placement `tff.AGGREGATORS` in order to maintain closure (as described later in the section).

Reply:
I'd actually suggest that (in the code you linked) the check should be pushed into the constructor of `federated_map`. Which it looks like it already is... https://github.com/tensorflow/federated/blob/c671f0b8857a3be18208db7a1c460f51650a7c60/tensorflow_federated/python/core/impl/intrinsic_factory.py#L177